### PR TITLE
Use FastRandom for catch droplets and bananas positions

### LINKED ISSUE #2232: Catch needs to correctly use FastRandom to offset droplets
Requires some more foundation around random seeding per hitobject.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/56182325-catch-needs-to-correctly-use-fastrandom-to-offset-droplets?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F4707269&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ekrctb**
Prereqs:
- [x] https://github.com/ppy/osu/pull/2806

Fixes #2232.
CatchBeatmapConversionTest is still not working due to several things.
HR is also need to use FastRandom but it is not in this PR.

### CODE REVIEW

**Comment by ekrctb** [comment]
Discarding 3 times is correct for spinners I believe.
Stable: screenshot473
Discarding 3 times: osu_2018-06-13_18-57-32
Discarding 2 times: osu_2018-06-13_19-03-00
[nekodex - circles! (ekr) [test].osu.txt](https://github.com/ppy/osu/files/2097665/nekodex.-.circles.ekr.test.osu.txt)

**Comment by smoogipoo** [comment]
@ekrctb Nice catch, this was an osu!stable-side issue that resulted in an incorrect output for my test cases.

**Comment by smoogipoo** [comment]
@peppy Should be ready for review now, this gets a 👍 from me, but I've made a few changes.

Note: Tests are currently ignored due to the failing "basic" case which will be fixed with the aforementioned slider changes (future PR).

**Comment by ekrctb** [comment]
With clamping and #2794, the remaining discrepancy is due to floating point number issue on slider conversion (stable always truncate time to int afaik).
I can think the solution will be one of the followings:
1. Don't mind the difference as it doesn't affect gameplay in a "major" way (it is debatable tho).
2. Try to match stable for opportunistic basis.
3. Try to match stable as precise as possible. It can be done given the stable source although the things of floating point numbers can be fairly tricky to cope with.
I note that the stable already changed (I don't know if it is intentional or not) the droplet generation as can be seen in maps for instance https://osu.ppy.sh/b/462700&m=2 (maximum combo is changed).

**Comment by ekrctb** [comment]
Also can someone have access to stable source check whether clamping is done in this way? (clamping is done after droplet sliding or before or both).

**Comment by smoogipoo** [comment]
Yup, clamping is done this way on stable.

It seems like the remaining discrepancy isn't due to time, but due to differences in curve calculations. They seem pretty big though so I'll look into it a bit.

**Comment by ekrctb** [comment]
@smoogipoo I'm not sure if/why we are getting different result but when I put Math.floor for some time 
values arbitrary (because I don't know proper way to doing so) on slider conversion, it passed "374472" case above (with absolute allowance 2). It generated without too many droplets without floor.

**Review by peppy — DISMISSED** [review]

**Review by peppy — APPROVED** [review]

File: osu.Game/Beatmaps/WorkingBeatmap.cs [hunk 1]
```diff
@@ -127,6 +124,9 @@ public IBeatmap GetPlayableBeatmap(RulesetInfo ruleset)
             foreach (var obj in converted.HitObjects)
                 mod.ApplyToHitObject(obj);
 
+            // Post-process
+            rulesetInstance.CreateBeatmapProcessor(converted)?.PostProcess();
```
Comment (peppy) [inline osu.Game/Beatmaps/WorkingBeatmap.cs]: this change not only caused a regression in catch behaviour (which i addressed) but also osu! and potentially other modes.

these kinds of changes need to be well tested and well documented going forwards.

image

**Pull request merged**